src/routers/alpha-router/functions/get-candidate-pools.ts

```diff
@@ -761,124 +733,225 @@ export async function getV2CandidatePools({
+          (subgraphPool.token1.id == wethAddress && subgraphPool.token0.id == tokenInAddress)
+        )
+      )
+    ) {
+      topByEthQuoteTokenPool.push(subgraphPool);
+      continue;
+    }
+
+    if (topByTVL.length < topN) {
+      topByTVL.push(subgraphPool);
+      continue;
+    }
+
+    if (
+      topByTVLUsingTokenIn.length < topNTokenInOut &&
+      (subgraphPool.token0.id == tokenInAddress || subgraphPool.token1.id == tokenInAddress)
+    ) {
+      topByTVLUsingTokenIn.push(subgraphPool);
+      continue;
+    }
+
+    if (
+      topByTVLUsingTokenOut.length < topNTokenInOut &&
+      (subgraphPool.token0.id == tokenOutAddress || subgraphPool.token1.id == tokenOutAddress)
+    ) {
+      topByTVLUsingTokenOut.push(subgraphPool);
+      continue;
+    }
   }
 
+  // Add the addresses found so far to our address set
+  addToAddressSet(topByBaseWithTokenIn);
+  addToAddressSet(topByBaseWithTokenOut);
   addToAddressSet(topByEthQuoteTokenPool);
+  addToAddressSet(topByTVLUsingTokenIn);
+  addToAddressSet(topByTVLUsingTokenOut);
+  addToAddressSet(topByTVL);
 
-  const topByTVL = _(subgraphPoolsSorted)
-    .filter((subgraphPool) => {
-      return !poolAddressesSoFar.has(subgraphPool.id);
-    })
-    .slice(0, topN)
-    .value();
+  // Filtering step for second hops
+  const topByTVLUsingTokenInSecondHopsMap: Map<string, V2SubgraphPool[]> = new Map();
+  const topByTVLUsingTokenOutSecondHopsMap: Map<string, V2SubgraphPool[]> = new Map();
+  const tokenInSecondHopAddresses = topByTVLUsingTokenIn.map((pool) =>
+    tokenInAddress == pool.token0.id ? pool.token1.id : pool.token0.id
+  );
+  const tokenOutSecondHopAddresses = topByTVLUsingTokenOut.map((pool) =>
+    tokenOutAddress == pool.token0.id ? pool.token1.id : pool.token0.id
+  );
 
-  addToAddressSet(topByTVL);
+  for (const secondHopId of tokenInSecondHopAddresses) {
+    topByTVLUsingTokenInSecondHopsMap.set(secondHopId, []);
+  }
+  for (const secondHopId of tokenOutSecondHopAddresses) {
+    topByTVLUsingTokenOutSecondHopsMap.set(secondHopId, []);
+  }
 
-  const topByTVLUsingTokenIn = _(subgraphPoolsSorted)
-    .filter((subgraphPool) => {
-      return (
-        !poolAddressesSoFar.has(subgraphPool.id) &&
-        (subgraphPool.token0.id == tokenInAddress ||
-          subgraphPool.token1.id == tokenInAddress)
-      );
-    })
-    .slice(0, topNTokenInOut)
-    .value();
+  for (const subgraphPool of subgraphPoolsSorted) {
+    if (poolAddressesSoFar.has(subgraphPool.id)) {
+      continue;
+    }
 
-  addToAddressSet(topByTVLUsingTokenIn);
+    // Only consider pools where neither tokens are in the blocked token list.
+    if (blockedTokenListProvider) {
+      const [token0InBlocklist, token1InBlocklist] = await Promise.all([
+        blockedTokenListProvider.getTokenByAddress(subgraphPool.token0.id),
+        blockedTokenListProvider.getTokenByAddress(subgraphPool.token1.id)
+      ]);
```

Comment:
hmm i know we're re-iterating over the subgraphPoolsSorted here but is there any way to not have to call the blockedTokenListProvider again since we already do it above?

Reply:
so I thought about this too, and I thought of collecting a list of results, however, the blockedTokenListProvider already contains a caching layer on top, so this is not really a big problem.